File: crates/spur-notebook/jute-notebook/src-tauri/src/assets/ports_bootstrap.go

!*go get github.com/apache/arrow-go/v18
!*go get github.com/janpfeifer/gonb/gonbui

// --- SPUR port helper bootstrap ---
import "encoding/json"
import "errors"
import "fmt"
import "html"
import "os"
import "path/filepath"
import "strings"

import "github.com/apache/arrow-go/v18/arrow"
import "github.com/apache/arrow-go/v18/arrow/ipc"
import "github.com/janpfeifer/gonb/gonbui"
import "github.com/janpfeifer/gonb/gonbui/protocol"

const portMime = "application/vnd.spur.port+json"
const portsDirName = "ports"
const portFileVersionSeparator = "@v"
const portArrowFileExtension = "arrow"
const portManifestFileName = "manifest.json"
const portManifestTempPrefix = "manifest.json."
const portTempFileSuffix = ".tmp"
const portManifestPortsKey = "ports"
const portManifestPathKey = "path"
const portManifestVersionKey = "version"
const portManifestSchemaKey = "schema"
const portInitialVersion uint64 = 0
const portVersionIncrement uint64 = 1
const portReservedCurrentDir = "."
const portReservedParentDir = ".."
const portForbiddenSlash = "/"
const portForbiddenBackslash = "\\"
const portForbiddenNul = "\u0000"

type spurPorts struct {
    root string
    portsDir string
    manifestPath string
    mime string
}

type spurManifest struct {
    Ports map[string]spurManifestEntry `json:"ports"`
}

type spurManifestEntry struct {
    Path string `json:"path"`
    Version uint64 `json:"version"`
    Schema any `json:"schema"`
}

func newSpurPorts(root, portsDir, manifestPath, mime string) *spurPorts {
    _ = os.MkdirAll(portsDir, 0o755)
    return &spurPorts{
        root: root,
        portsDir: portsDir,
        manifestPath: manifestPath,
        mime: mime,
    }
}

func (s *spurPorts) Put(port string, batch arrow.Record) (map[string]any, error) {
    if batch == nil {
        return nil, errors.New("SPUR port batch cannot be nil")
    }
    if err := s.validatePort(port); err != nil {
        return nil, err
    }
    if err := os.MkdirAll(s.portsDir, 0o755); err != nil {
        return nil, err
    }

    manifest, err := s.loadManifest()
    if err != nil {
        return nil, err
    }
    schema, err := spurSchemaJSON(batch.Schema(), port)
    if err != nil {
        return nil, err
    }
    previousVersion := portInitialVersion
    if entry, ok := manifest.Ports[port]; ok {
        previousVersion = entry.Version
    }
    version := previousVersion + portVersionIncrement
    arrowPath := filepath.Join(
        s.portsDir,
        fmt.Sprintf("%s%s%d.%s", port, portFileVersionSeparator, version, portArrowFileExtension),
    )

    if err := s.writeArrowFile(arrowPath, batch); err != nil {
        return nil, err
    }

    manifest.Ports[port] = spurManifestEntry{
        Path: arrowPath,
        Version: version,
        Schema: schema,
    }
    if err := s.storeManifest(manifest); err != nil {
        return nil, err
    }

    payload := map[string]any{
        "port": port,
        portManifestVersionKey: version,
        portManifestSchemaKey: schema,
    }
    s.displayPort(port, version, schema, batch)
    return payload, nil
}

func (s *spurPorts) Get(port string) ([]arrow.Record, error) {
    if err := s.validatePort(port); err != nil {
        return nil, err
    }
    manifest, err := s.loadManifest()
    if err != nil {
        return nil, err
    }
    entry, ok := manifest.Ports[port]
    if !ok {
        return nil, fmt.Errorf("SPUR port has not been written: %s", port)
    }

    file, err := os.Open(entry.Path)
    if err != nil {
        return nil, err
    }
    defer file.Close()
    reader, err := ipc.NewFileReader(file)
    if err != nil {
        return nil, err
    }
    defer reader.Close()

    records := make([]arrow.Record, 0, reader.NumRecords())
    for i := 0; i < reader.NumRecords(); i++ {
        record, err := reader.RecordAt(i)
        if err != nil {
            return nil, err
        }
        records = append(records, record)
    }
    return records, nil
}

func (s *spurPorts) writeArrowFile(path string, batch arrow.Record) error {
    file, err := os.Create(path)
    if err != nil {
        return err
    }
    writer, err := ipc.NewFileWriter(file, ipc.WithSchema(batch.Schema()))
    if err != nil {
        _ = file.Close()
        return err
    }
    if err := writer.Write(batch); err != nil {
        _ = writer.Close()
        _ = file.Close()
        return err
    }
    if err := writer.Close(); err != nil {
        _ = file.Close()
        return err
    }
    return file.Close()
}

func (s *spurPorts) loadManifest() (spurManifest, error) {
    manifest := spurManifest{Ports: map[string]spurManifestEntry{}}
    data, err := os.ReadFile(s.manifestPath)
    if errors.Is(err, os.ErrNotExist) {
        return manifest, nil
    }
    if err != nil {
        return manifest, err
    }
    if err := json.Unmarshal(data, &manifest); err != nil {
        return manifest, err
    }
    if manifest.Ports == nil {
        manifest.Ports = map[string]spurManifestEntry{}
    }
    return manifest, nil
}

func (s *spurPorts) storeManifest(manifest spurManifest) error {
    tmp, err := os.CreateTemp(
        s.portsDir,
        fmt.Sprintf("%s*%s", portManifestTempPrefix, portTempFileSuffix),
    )
    if err != nil {
        return err
    }
    tmpPath := tmp.Name()
    cleanup := true
    defer func() {
        if cleanup {
            _ = os.Remove(tmpPath)
        }
    }()

    encoder := json.NewEncoder(tmp)
    encoder.SetIndent("", "  ")
    if err := encoder.Encode(manifest); err != nil {
        _ = tmp.Close()
        return err
    }
    if err := tmp.Close(); err != nil {
        return err
    }
    if err := os.Rename(tmpPath, s.manifestPath); err != nil {
        return err
    }
    cleanup = false
    return nil
}

func (s *spurPorts) validatePort(port string) error {
    if port == "" {
        return errors.New("SPUR port name cannot be empty")
    }
    if port == portReservedCurrentDir ||
        port == portReservedParentDir ||
        strings.Contains(port, portForbiddenSlash) ||
        strings.Contains(port, portForbiddenBackslash) ||
        strings.Contains(port, portForbiddenNul) {
        return fmt.Errorf("SPUR port name is not valid for an on-disk port file: %s", port)
    }
    return nil
}

func (s *spurPorts) displayPort(port string, version uint64, schema map[string]any, batch arrow.Record) {
    gonbui.SendData(&protocol.DisplayData{
        Data: map[protocol.MIMEType]any{
            protocol.MIMEType(s.mime): map[string]any{
                "port": port,
                portManifestVersionKey: version,
                portManifestSchemaKey: schema,
            },
            protocol.MIMETextHTML: s.previewHTML(port, version, batch),
        },
    })
    gonbui.Sync()
}

func (s *spurPorts) previewHTML(port string, version uint64, batch arrow.Record) string {
    return fmt.Sprintf(
        "<div><strong>SPUR port</strong> <code>%s</code> <span>v%d</span><p>%d rows x %d columns</p></div>",
        html.EscapeString(port),
        version,
        batch.NumRows(),
        batch.NumCols(),
    )
}

func spurSchemaJSON(schema *arrow.Schema, port string) (map[string]any, error) {
    if schema == nil {
        return nil, fmt.Errorf("SPUR port %q: Arrow schema cannot be nil", port)
    }
    fields := make([]map[string]any, 0, schema.NumFields())
    for _, field := range schema.Fields() {
        dataType, err := spurDataTypeJSON(field.Type, port)
        if err != nil {
            return nil, err
        }
        fields = append(fields, map[string]any{
            "name": field.Name,
            "data_type": dataType,
            "nullable": field.Nullable,
            "dict_id": 0,
            "dict_is_ordered": false,
            "metadata": spurMetadataJSON(field.Metadata),
        })
    }
    return map[string]any{
        "fields": fields,
        "metadata": spurMetadataJSON(schema.Metadata()),
    }, nil
}

func spurMetadataJSON(metadata arrow.Metadata) map[string]string {
    output := map[string]string{}
    for key, value := range metadata.ToMap() {
        output[key] = value
    }
    return output
}

func spurDataTypeJSON(dataType arrow.DataType, port string) (any, error) {
    switch t := dataType.(type) {
    case *arrow.BooleanType:
        return "Boolean", nil
    case *arrow.Int8Type:
        return "Int8", nil
    case *arrow.Int16Type:
        return "Int16", nil
    case *arrow.Int32Type:
        return "Int32", nil
    case *arrow.Int64Type:
        return "Int64", nil
    case *arrow.Uint8Type:
        return "UInt8", nil
    case *arrow.Uint16Type:
        return "UInt16", nil
    case *arrow.Uint32Type:
        return "UInt32", nil
    case *arrow.Uint64Type:
        return "UInt64", nil
    case *arrow.Float16Type:
        return "Float16", nil
    case *arrow.Float32Type:
        return "Float32", nil
    case *arrow.Float64Type:
        return "Float64", nil
    case *arrow.StringType:
        return "Utf8", nil
    case *arrow.LargeStringType:
        return "LargeUtf8", nil
    case *arrow.BinaryType:
        return "Binary", nil
    case *arrow.LargeBinaryType:
        return "LargeBinary", nil
    case *arrow.NullType:
        return "Null", nil
    case *arrow.Date32Type:
        return "Date32", nil
    case *arrow.Date64Type:
        return "Date64", nil
    case *arrow.TimestampType:
        unit, err := spurTimeUnitJSON(t.Unit, port, dataType)
        if err != nil {
            return nil, err
        }
        var timezone any
        if t.TimeZone != "" {
            timezone = t.TimeZone
        }
        return map[string]any{"Timestamp": []any{unit, timezone}}, nil
    case *arrow.Time32Type:
        unit, err := spurTimeUnitJSON(t.Unit, port, dataType)
        if err != nil {
            return nil, err
        }
        return map[string]any{"Time32": unit}, nil
    case *arrow.Time64Type:
        unit, err := spurTimeUnitJSON(t.Unit, port, dataType)
        if err != nil {
            return nil, err
        }
        return map[string]any{"Time64": unit}, nil
    case *arrow.Decimal128Type:
        return map[string]any{"Decimal128": []any{t.GetPrecision(), t.GetScale()}}, nil
    case *arrow.Decimal256Type:
        return map[string]any{"Decimal256": []any{t.GetPrecision(), t.GetScale()}}, nil
    case *arrow.DictionaryType:
        index, err := spurDataTypeJSON(t.IndexType, port)
        if err != nil {
            return nil, err
        }
        value, err := spurDataTypeJSON(t.ValueType, port)
        if err != nil {
            return nil, err
        }
        return map[string]any{"Dictionary": []any{index, value}}, nil
    default:
        return nil, fmt.Errorf("SPUR port %q: unsupported Arrow type for manifest schema: %s", port, dataType)
    }
}

func spurTimeUnitJSON(unit arrow.TimeUnit, port string, dataType arrow.DataType) (string, error) {
    switch unit {
    case arrow.Second:
        return "Second", nil
    case arrow.Millisecond:
        return "Millisecond", nil
    case arrow.Microsecond:
        return "Microsecond", nil
    case arrow.Nanosecond:
        return "Nanosecond", nil
    default:
        return "", fmt.Errorf("SPUR port %q: unsupported Arrow time unit for manifest schema: %s", port, dataType)
    }
}

var spurPortRoot = func() string {
    root := os.Getenv("SPUR_NOTEBOOK_PORT_ROOT")
    if root == "" {
        panic("SPUR_NOTEBOOK_PORT_ROOT is not set")
    }
    return root
}()
var spur = newSpurPorts(
    spurPortRoot,
    filepath.Join(spurPortRoot, portsDirName),
    filepath.Join(spurPortRoot, portsDirName, portManifestFileName),
    portMime,
)
// --- end SPUR port helper bootstrap ---

File: crates/spur-notebook/jute-notebook/src-tauri/src/assets/ports_bootstrap.rs
:dep arrow = "55"
:dep serde_json = "1"

#[derive(Clone, Debug)]
struct _Spur {
    root: String,
}

use std::fs;
use std::fs::File;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use arrow::ipc::reader::FileReader;
use arrow::ipc::writer::FileWriter;
use arrow::record_batch::RecordBatch;
use serde_json::{Map, Number, Value};

const PORT_MIME: &str = "application/vnd.spur.port+json";
const PORTS_DIR_NAME: &str = "ports";
const PORT_FILE_VERSION_SEPARATOR: &str = "@v";
const PORT_ARROW_FILE_EXTENSION: &str = "arrow";
const PORT_MANIFEST_FILE_NAME: &str = "manifest.json";
const PORT_MANIFEST_TEMP_PREFIX: &str = "manifest.json.";
const PORT_TEMP_FILE_SUFFIX: &str = ".tmp";
const PORT_MANIFEST_PORTS_KEY: &str = "ports";
const PORT_MANIFEST_PATH_KEY: &str = "path";
const PORT_MANIFEST_VERSION_KEY: &str = "version";
const PORT_MANIFEST_SCHEMA_KEY: &str = "schema";
const PORT_INITIAL_VERSION: u64 = 0;
const PORT_VERSION_INCREMENT: u64 = 1;
const PORT_RESERVED_CURRENT_DIR: &str = ".";
const PORT_RESERVED_PARENT_DIR: &str = "..";
const PORT_FORBIDDEN_SLASH: &str = "/";
const PORT_FORBIDDEN_BACKSLASH: &str = "\\";
const PORT_FORBIDDEN_NUL: &str = "\0";

impl _Spur {
    pub fn new(root: impl Into<String>) -> Self {
        Self { root: root.into() }
    }

    pub fn put(
        &self,
        port: &str,
        batch: RecordBatch,
    ) -> Result<Value, Box<dyn std::error::Error>> {
        self.validate_port(port)?;
        let ports_dir = self.ports_dir();
        fs::create_dir_all(&ports_dir)?;

        let mut manifest = self.load_manifest()?;
        let schema = Self::schema_json(batch.schema().as_ref(), port)?;
        let previous_version = Self::previous_version(&manifest, port);
        let version = previous_version + PORT_VERSION_INCREMENT;
        let arrow_path = ports_dir.join(Self::port_file_name(port, version));

        let file = File::create(&arrow_path)?;
        let mut writer = FileWriter::try_new(file, batch.schema().as_ref())?;
        writer.write(&batch)?;
        writer.finish()?;

        Self::set_manifest_entry(
            &mut manifest,
            port,
            arrow_path.to_string_lossy().into_owned(),
            version,
            schema.clone(),
        )?;
        self.store_manifest(&manifest)?;
        self.display_port(port, version, schema, &batch)
    }

    pub fn get(
        &self,
        port: &str,
    ) -> Result<Vec<RecordBatch>, Box<dyn std::error::Error>> {
        self.validate_port(port)?;
        let manifest = self.load_manifest()?;
        let entry = manifest
            .get(PORT_MANIFEST_PORTS_KEY)
            .and_then(|ports| ports.get(port))
            .ok_or_else(|| {
                std::io::Error::new(
                    std::io::ErrorKind::NotFound,
                    format!("SPUR port has not been written: {port}"),
                )
            })?;
        let path = entry
            .get(PORT_MANIFEST_PATH_KEY)
            .and_then(Value::as_str)
            .ok_or_else(|| {
                std::io::Error::new(
                    std::io::ErrorKind::InvalidData,
                    format!("SPUR port manifest entry is missing a path: {port}"),
                )
            })?;

        let file = File::open(path)?;
        let reader = FileReader::try_new(file, None)?;
        let mut batches = Vec::new();
        for batch in reader {
            batches.push(batch?);
        }
        Ok(batches)
    }

    fn ports_dir(&self) -> PathBuf {
        Path::new(self.root.as_str()).join(PORTS_DIR_NAME)
    }

    fn manifest_path(&self) -> PathBuf {
        self.ports_dir().join(PORT_MANIFEST_FILE_NAME)
    }

    fn load_manifest(&self) -> Result<Value, Box<dyn std::error::Error>> {
        let manifest_path = self.manifest_path();
        if !manifest_path.exists() {
            return Ok(Self::empty_manifest());
        }
        let text = fs::read_to_string(manifest_path)?;
        Ok(serde_json::from_str(&text)?)
    }

    fn store_manifest(&self, manifest: &Value) -> Result<(), Box<dyn std::error::Error>> {
        let tmp_path = self.temp_manifest_path()?;
        let mut tmp = File::create(&tmp_path)?;
        serde_json::to_writer_pretty(&mut tmp, manifest)?;
        writeln!(&mut tmp)?;
        drop(tmp);
        fs::rename(&tmp_path, self.manifest_path())?;
        Ok(())
    }

    fn temp_manifest_path(&self) -> Result<PathBuf, Box<dyn std::error::Error>> {
        let nanos = SystemTime::now().duration_since(UNIX_EPOCH)?.as_nanos();
        Ok(self.ports_dir().join(format!(
            "{PORT_MANIFEST_TEMP_PREFIX}{nanos}{PORT_TEMP_FILE_SUFFIX}"
        )))
    }

    fn previous_version(manifest: &Value, port: &str) -> u64 {
        manifest
            .get(PORT_MANIFEST_PORTS_KEY)
            .and_then(|ports| ports.get(port))
            .and_then(|entry| entry.get(PORT_MANIFEST_VERSION_KEY))
            .and_then(Value::as_u64)
            .unwrap_or(PORT_INITIAL_VERSION)
    }

    fn port_file_name(port: &str, version: u64) -> String {
        format!(
            "{}{}{}.{}",
            port, PORT_FILE_VERSION_SEPARATOR, version, PORT_ARROW_FILE_EXTENSION
        )
    }

    fn empty_manifest() -> Value {
        let mut manifest = Map::new();
        manifest.insert(
            PORT_MANIFEST_PORTS_KEY.to_string(),
            Value::Object(Map::new()),
        );
        Value::Object(manifest)
    }

    fn set_manifest_entry(
        manifest: &mut Value,
        port: &str,
        path: String,
        version: u64,
        schema: Value,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let manifest_object = manifest.as_object_mut().ok_or_else(|| {
            std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "SPUR port manifest must be a JSON object",
            )
        })?;
        let ports_value = manifest_object
            .entry(PORT_MANIFEST_PORTS_KEY.to_string())
            .or_insert_with(|| Value::Object(Map::new()));
        let ports = ports_value.as_object_mut().ok_or_else(|| {
            std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "SPUR port manifest ports must be a JSON object",
            )
        })?;

        let mut entry = Map::new();
        entry.insert(PORT_MANIFEST_PATH_KEY.to_string(), Value::String(path));
        entry.insert(
            PORT_MANIFEST_VERSION_KEY.to_string(),
            Value::Number(Number::from(version)),
        );
        entry.insert(PORT_MANIFEST_SCHEMA_KEY.to_string(), schema);
        ports.insert(port.to_string(), Value::Object(entry));
        Ok(())
    }

    fn schema_json(
        schema: &arrow::datatypes::Schema,
        port: &str,
    ) -> Result<Value, Box<dyn std::error::Error>> {
        let mut fields = Vec::new();
        for field in schema.fields() {
            let mut field_json = Map::new();
            field_json.insert("name".to_string(), Value::String(field.name().clone()));
            field_json.insert(
                "data_type".to_string(),
                Self::data_type_json(field.data_type(), port)?,
            );
            field_json.insert("nullable".to_string(), Value::Bool(field.is_nullable()));
            field_json.insert("dict_id".to_string(), Value::Number(Number::from(0)));
            field_json.insert("dict_is_ordered".to_string(), Value::Bool(false));
            field_json.insert("metadata".to_string(), Self::metadata_json(field.metadata()));
            fields.push(Value::Object(field_json));
        }

        let mut schema_json = Map::new();
        schema_json.insert("fields".to_string(), Value::Array(fields));
        schema_json.insert("metadata".to_string(), Self::metadata_json(schema.metadata()));
        Ok(Value::Object(schema_json))
    }

    fn metadata_json(metadata: &std::collections::HashMap<String, String>) -> Value {
        let mut output = Map::new();
        for (key, value) in metadata {
            output.insert(key.clone(), Value::String(value.clone()));
        }
        Value::Object(output)
    }

    fn data_type_json(
        data_type: &arrow::datatypes::DataType,
        port: &str,
    ) -> Result<Value, Box<dyn std::error::Error>> {
        use arrow::datatypes::DataType;

        let value = match data_type {
            DataType::Null => "Null",
            DataType::Boolean => "Boolean",
            DataType::Int8 => "Int8",
            DataType::Int16 => "Int16",
            DataType::Int32 => "Int32",
            DataType::Int64 => "Int64",
            DataType::UInt8 => "UInt8",
            DataType::UInt16 => "UInt16",
            DataType::UInt32 => "UInt32",
            DataType::UInt64 => "UInt64",
            DataType::Float16 => "Float16",
            DataType::Float32 => "Float32",
            DataType::Float64 => "Float64",
            DataType::Utf8 => "Utf8",
            DataType::LargeUtf8 => "LargeUtf8",
            DataType::Binary => "Binary",
            DataType::LargeBinary => "LargeBinary",
            DataType::Date32 => "Date32",
            DataType::Date64 => "Date64",
            other => {
                return Err(std::io::Error::new(
                    std::io::ErrorKind::InvalidInput,
                    format!(
                        "SPUR port {port:?}: unsupported Arrow type for manifest schema: {other}"
                    ),
                )
                .into())
            }
        };
        Ok(Value::String(value.to_string()))
    }

    fn display_port(
        &self,
        port: &str,
        version: u64,
        schema: Value,
        batch: &RecordBatch,
    ) -> Result<Value, Box<dyn std::error::Error>> {
        let mut payload = Map::new();
        payload.insert("port".to_string(), Value::String(port.to_string()));
        payload.insert(
            PORT_MANIFEST_VERSION_KEY.to_string(),
            Value::Number(Number::from(version)),
        );
        payload.insert(PORT_MANIFEST_SCHEMA_KEY.to_string(), schema);
        let payload = Value::Object(payload);
        let payload_json = serde_json::to_string(&payload)?;
        println!(
            "EVCXR_BEGIN_CONTENT {}\n{}\nEVCXR_END_CONTENT",
            PORT_MIME, payload_json
        );
        println!(
            "EVCXR_BEGIN_CONTENT text/html\n{}\nEVCXR_END_CONTENT",
            self.preview_html(port, version, batch)
        );
        Ok(payload)
    }

    fn preview_html(&self, port: &str, version: u64, batch: &RecordBatch) -> String {
        let title = format!(
            "<strong>SPUR port</strong> <code>{}</code> <span>v{}</span>",
            Self::escape_html(port),
            version
        );
        format!(
            "<div>{title}<p>{} rows x {} columns</p></div>",
            batch.num_rows(),
            batch.num_columns()
        )
    }

    fn escape_html(value: &str) -> String {
        value
            .replace('&', "&amp;")
            .replace('<', "&lt;")
            .replace('>', "&gt;")
            .replace('"', "&quot;")
            .replace('\'', "&#39;")
    }

    fn validate_port(&self, port: &str) -> Result<(), Box<dyn std::error::Error>> {
        if port.is_empty() {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "SPUR port name cannot be empty",
            )
            .into());
        }
        if port == PORT_RESERVED_CURRENT_DIR
            || port == PORT_RESERVED_PARENT_DIR
            || port.contains(PORT_FORBIDDEN_SLASH)
            || port.contains(PORT_FORBIDDEN_BACKSLASH)
            || port.contains(PORT_FORBIDDEN_NUL)
        {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                format!("SPUR port name is not valid for an on-disk port file: {port}"),
            )
            .into());
        }
        Ok(())
    }
}

let spur: _Spur = _Spur::new(std::env::var("SPUR_NOTEBOOK_PORT_ROOT").expect("SPUR_NOTEBOOK_PORT_ROOT is not set"));
